File: frontend/src/lib/api/classes.ts
// Classes/Programs API service — uses Supabase directly (table: teams)
// Per NAMING_DICTIONARY.md: "classes" in UI = "teams" in Supabase
import { supabase } from '@/integrations/supabase/client';
import { bffClient } from './bffClient';
type TeamRow = any;
type TeamWithCoach = TeamRow & { coach?: { full_name: string | null } | null };

export interface Schedule {
  day: 'monday' | 'tuesday' | 'wednesday' | 'thursday' | 'friday' | 'saturday' | 'sunday';
  start_time: string;
  end_time: string;
}

const isValidUUID = (id: string | null | undefined): boolean => {
  if (!id) return false;
  const regex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
  return regex.test(id);
};

export interface Class {
  id: string;
  name: string;
  description?: string;
  sport: string;
  level: 'beginner' | 'intermediate' | 'advanced';
  school_id: string;
  coach_id?: string;
  coach_name?: string;
  capacity: number;
  enrolled_count: number;
  current_participants?: number;
  schedule: Schedule[];
  location?: string;
  price?: number;
  price_monthly?: number;
  status: 'active' | 'inactive' | 'full' | 'cancelled';
  start_date?: string;
  end_date?: string;
  active?: boolean;
  created_at: string;
  updated_at: string;
}

export interface ClassCreate {
  name: string;
  description?: string;
  sport: string;
  level?: 'beginner' | 'intermediate' | 'advanced';
  school_id: string;
  coach_id?: string;
  coach_name?: string;
  capacity?: number;
  schedule?: Schedule[];
  location?: string;
  price?: number;
  status?: 'active' | 'inactive' | 'full' | 'cancelled';
  start_date?: string;
  end_date?: string;
}

export interface ClassUpdate {
  name?: string;
  description?: string;
  sport?: string;
  level?: 'beginner' | 'intermediate' | 'advanced';
  coach_id?: string;
  coach_name?: string;
  capacity?: number;
  schedule?: Schedule[];
  location?: string;
  price?: number;
  status?: 'active' | 'inactive' | 'full' | 'cancelled';
  start_date?: string;
  end_date?: string;
}

export interface EnrollmentRecord {
  id: string;
  class_id: string;
  student_id: string;
  student_name: string;
  enrollment_date: string;
  status: 'active' | 'dropped' | 'completed';
}

export interface ClassStats {
  total: number;
  active: number;
  full: number;
  by_sport: Record<string, number>;
  total_enrolled: number;
}

class ClassesAPI {
  /**
   * Create a new program (class) in Supabase
   */
  async createClass(classData: ClassCreate): Promise<Class> {
    const { data, error } = await supabase
      .from('teams')
      .insert({
        name: classData.name,
        description: classData.description,
        sport: classData.sport,
        school_id: classData.school_id,
        coach_id: classData.coach_id,
        price_monthly: classData.price || 0,
        max_students: classData.capacity || 20,
        active: classData.status !== 'inactive',
      })
      .select(`
        *,
        coach:coach_id(full_name)
      `)
      .single();

    if (error) {
      console.error('Error creating program:', error);
      throw new Error(error.message || 'Failed to create class');
    }

    return this.mapTeamToClass(data);
  }

  /**
   * Get programs (classes) with filters
   */
  async getClasses(params?: {
    school_id?: string;
    sport?: string;
    level?: string;
    status?: string;
    coach_id?: string;
    search?: string;
    skip?: number;
    limit?: number;
  }): Promise<Class[]> {
    try {
      if (!params?.school_id || !isValidUUID(params.school_id)) {
        console.warn('getClasses called without valid school_id, returning empty array');
        return [];
      }

      let query = supabase
        .from('teams')
        .select(`
          *,
          coach:coach_id(full_name)
        `)
        .eq('school_id', params.school_id)
        .order('created_at', { ascending: false });
      if (params?.sport) {
        query = query.eq('sport', params.sport);
      }
      if (params?.status === 'active') {
        query = query.eq('active', true);
      } else if (params?.status === 'inactive') {
        query = query.eq('active', false);
      }
      if (params?.search) {
        query = query.ilike('name', `%${params.search}%`);
      }

      const offset = params?.skip ?? 0;
      const limit = params?.limit ?? 100;
      query = query.range(offset, offset + limit - 1);

      const { data, error } = await query;

      if (error) {
        console.warn('Error fetching programs from Supabase:', error.message);
        return [];
      }

      return (data || []).map((t) => this.mapTeamToClass(t as TeamWithCoach));
    } catch (error) {
      console.warn('Error fetching classes:', error);
      return [];
    }
  }

  /**
   * Get a single program by ID
   */
  async getClass(id: string): Promise<Class> {
    const { data, error } = await supabase
      .from('teams')
      .select(`
        *,
        coach:coach_id(full_name)
      `)
      .eq('id', id)
      .single();

    if (error || !data) {
      throw new Error(error?.message || 'Class not found');
    }

    return this.mapTeamToClass(data);
  }

  /**
   * Update a program
   */
  async updateClass(id: string, updates: ClassUpdate): Promise<Class> {
    const updateData: any = {};
    if (updates.name !== undefined) updateData.name = updates.name;
    if (updates.description !== undefined) updateData.description = updates.description;
    if (updates.sport !== undefined) updateData.sport = updates.sport;
    if (updates.price !== undefined) updateData.price_monthly = updates.price;
    if (updates.capacity !== undefined) updateData.max_students = updates.capacity;
    if (updates.status !== undefined) updateData.active = updates.status !== 'inactive';
    if (updates.coach_id !== undefined) updateData.coach_id = updates.coach_id;
    updateData.updated_at = new Date().toISOString();

    const { data, error } = await supabase
      .from('teams')
      .update(updateData)
      .eq('id', id)
      .select(`
        *,
        coach:coach_id(full_name)
      `)
      .single();

    if (error || !data) {
      throw new Error(error?.message || 'Failed to update class');
    }

    return this.mapTeamToClass(data);
  }

  /**
   * Delete a program
   */
  async deleteClass(id: string): Promise<void> {
    const { error } = await supabase
      .from('teams')
      .delete()
      .eq('id', id);

    if (error) {
      throw new Error(error.message || 'Failed to delete class');
    }
  }

  /**
   * Enroll a student in a program using BFF endpoint
   */
  async enrollStudent(classId: string, studentId: string, studentName: string, isAdult: boolean = false): Promise<EnrollmentRecord> {
    const payload = isAdult
      ? { user_id: studentId, team_id: classId }
      : { child_id: studentId, team_id: classId };

    const data = await bffClient.post<{ id: string; created_at: string }>(
      '/api/v1/enrollments',
      payload
    );

    return {
      id: data.id,
      class_id: classId,
      student_id: studentId,
      student_name: studentName,
      enrollment_date: data.created_at,
      status: 'active',
    };
  }

  /**
   * Unenroll a student from a program
   */
  async unenrollStudent(classId: string, studentId: string): Promise<void> {
    const { error } = await supabase
      .from('enrollments')
      .update({ status: 'cancelled' })
      .eq('team_id', classId)
      .eq('child_id', studentId)
      .eq('status', 'active');

    if (error) {
      console.warn('Error unenrolling student:', error.message);
    }
  }

  /**
   * Get all students enrolled in a program
   */
  async getClassStudents(classId: string): Promise<any[]> {
    try {
      // Obtener enrollments activos con user_id y child_id
      const { data: enrollments, error } = await supabase
        .from('enrollments')
        .select('id, created_at, child_id, user_id')
        .eq('team_id', classId)
        .eq('status', 'active');

      if (error) {
        console.warn('Error fetching enrolled students:', error.message);
        return [];
      }
      if (!enrollments || enrollments.length === 0) return [];

      // Recopilar IDs de ambos tipos
      const childIds = enrollments.map(e => e.child_id).filter(Boolean) as string[];
      const userIds = enrollments.map(e => e.user_id).filter(Boolean) as string[];
      const allIds = [...childIds, ...userIds];

      if (allIds.length === 0) return [];

      // Una sola query a school_athletes por ID
      const { data: athletes } = await (supabase as any)
        .from('school_athletes')
        .select('id, full_name, date_of_birth, avatar_url, athlete_type, parent_name')
        .in('id', allIds);

      const athleteMap = Object.fromEntries(
        (athletes || []).map((a: any) => [a.id, a])
      );

      return enrollments.map((enrollment: any) => {
        const athleteId = enrollment.child_id ?? enrollment.user_id;
        return {
          enrollment_id: enrollment.id,
          enrollment_date: enrollment.created_at,
          student: athleteMap[athleteId] ?? null,
        };
      });
    } catch (error) {
      console.warn('Error fetching class students:', error);
      return [];
    }
  }

  /**
   * Get class/program statistics for a school
   */
  async getStats(schoolId: string, branchId?: string | null, coachId?: string): Promise<ClassStats> {
    if ((!schoolId || !isValidUUID(schoolId)) && !coachId) {
      return { total: 0, active: 0, full: 0, by_sport: {}, total_enrolled: 0 };
    }
    try {
      let query = supabase
        .from('teams')
        .select('*');

      if (schoolId) {
        query = query.eq('school_id', schoolId);
      }

      if (branchId) {
        query = query.eq('branch_id', branchId);
      }

      if (coachId) {
        // Obtener equipos via campo legacy Y junction table
        const [{ data: legacyTeams }, { data: junctionTeams }] = await Promise.all([
          supabase.from('teams').select('id').eq('coach_id', coachId),
          supabase.from('team_coaches').select('team_id').eq('coach_id', coachId),
        ]);
        const teamIds = [...new Set([
          ...(legacyTeams || []).map(t => t.id),
          ...(junctionTeams || []).map(t => t.team_id),
        ])];
        if (teamIds.length === 0) {
          return { total: 0, active: 0, full: 0, by_sport: {}, total_enrolled: 0 };
        }
        query = query.in('id', teamIds);
      }

      const { data: teams } = await query;

      if (!teams) return { total: 0, active: 0, full: 0, by_sport: {}, total_enrolled: 0 };

      const total = teams.length;
      const active = teams.filter(t => t.active).length;
      const full = 0; // Simple logic, can refine if capacity/student count comparison is needed

      const by_sport: Record<string, number> = {};
      teams.forEach(t => {
        if (t.sport) {
          by_sport[t.sport] = (by_sport[t.sport] || 0) + 1;
        }
      });

      const total_enrolled = teams.reduce((sum, t) => sum + (t.current_students || 0), 0);

      return { total, active, full, by_sport, total_enrolled };
    } catch (error) {
      console.warn('Error fetching class stats:', error);
      return { total: 0, active: 0, full: 0, by_sport: {}, total_enrolled: 0 };
    }
  }

  /**
   * Map a Supabase 'teams' row to the Class interface
   */
  private mapTeamToClass(team: TeamWithCoach): Class {
    return {
      id: team.id,
      name: team.name,
      description: team.description,
      sport: team.sport || '',
      level: (team.level || team.age_group || 'beginner') as 'beginner' | 'intermediate' | 'advanced',
      school_id: team.school_id || '',
      coach_id: team.coach_id,
      coach_name: team.coach?.full_name,
      capacity: team.max_students || 20,
      enrolled_count: team.current_students || 0,
      current_participants: team.current_students || 0,
      schedule: (team.schedule as unknown as Schedule[]) || [],
      price: team.price_monthly,
      price_monthly: team.price_monthly,
      status: team.active ? 'active' : 'inactive',
      active: team.active,
      created_at: team.created_at,
      updated_at: team.updated_at,
    };
  }
}

export const classesAPI = new ClassesAPI();

File: backend/routes/classes.py
from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field, validator
from typing import List, Optional
from datetime import datetime, time
import uuid

router = APIRouter(prefix="/api/classes", tags=["classes"])

# MongoDB connection (will be injected)
db = None

def init_db(database):
    global db
    db = database

# Models
class Schedule(BaseModel):
    day: str = Field(..., pattern="^(monday|tuesday|wednesday|thursday|friday|saturday|sunday)$")
    start_time: str = Field(..., description="HH:MM format")
    end_time: str = Field(..., description="HH:MM format")
    
    @validator('start_time', 'end_time')
    def validate_time_format(cls, v):
        try:
            time.fromisoformat(v)
        except ValueError:
            raise ValueError('Time must be in HH:MM format')
        return v

class ClassBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    sport: str = Field(..., min_length=1, max_length=100)
    level: str = Field(..., pattern="^(beginner|intermediate|advanced)$")
    school_id: str
    coach_id: Optional[str] = None
    coach_name: Optional[str] = None
    capacity: int = Field(default=20, ge=1, le=100)
    schedule: List[Schedule] = Field(default_factory=list)
    location: Optional[str] = None
    price: Optional[float] = Field(None, ge=0)
    status: str = Field(default="active", pattern="^(active|inactive|full|cancelled)$")
    start_date: Optional[str] = None
    end_date: Optional[str] = None

class ClassCreate(ClassBase):
    pass

class ClassUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    sport: Optional[str] = None
    level: Optional[str] = None
    coach_id: Optional[str] = None
    coach_name: Optional[str] = None
    capacity: Optional[int] = Field(None, ge=1, le=100)
    schedule: Optional[List[Schedule]] = None
    location: Optional[str] = None
    price: Optional[float] = Field(None, ge=0)
    status: Optional[str] = None
    start_date: Optional[str] = None
    end_date: Optional[str] = None

class Class(ClassBase):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    enrolled_count: int = Field(default=0)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

class EnrollmentRecord(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    class_id: str
    student_id: str
    student_name: str
    enrollment_date: datetime = Field(default_factory=datetime.utcnow)
    status: str = Field(default="active", pattern="^(active|dropped|completed)$")

# Endpoints
@router.post("/", response_model=Class)
async def create_class(class_data: ClassCreate):
    """Create a new class/program"""
    try:
        class_dict = class_data.dict()
        class_obj = Class(**class_dict)
        
        await db.classes.insert_one(class_obj.dict())
        return class_obj
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))

@router.get("/", response_model=List[Class])
async def get_classes(
    school_id: Optional[str] = Query(None),
    sport: Optional[str] = Query(None),
    level: Optional[str] = Query(None),
    status: Optional[str] = Query(None),
    coach_id: Optional[str] = Query(None),
    search: Optional[str] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=500)
):
    """Get list of classes with filters"""
    try:
        query = {}
        
        if school_id:
            query['school_id'] = school_id
        if sport:
            query['sport'] = sport
        if level:
            query['level'] = level
        if status:
            query['status'] = status
        if coach_id:
            query['coach_id'] = coach_id
        if search:
            query['$or'] = [
                {'name': {'$regex': search, '$options': 'i'}},
                {'sport': {'$regex': search, '$options': 'i'}},
                {'coach_name': {'$regex': search, '$options': 'i'}}
            ]
        
        classes = await db.classes.find(query).skip(skip).limit(limit).to_list(limit)
        
        # Get enrollment count for each class
        for class_item in classes:
            count = await db.enrollments.count_documents({
                'class_id': class_item['id'],
                'status': 'active'
            })
            class_item['enrolled_count'] = count
        
        return [Class(**c) for c in classes]
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/{class_id}", response_model=Class)
async def get_class(class_id: str):
    """Get a specific class by ID"""
    try:
        class_item = await db.classes.find_one({"id": class_id})
        if not class_item:
            raise HTTPException(status_code=404, detail="Class not found")
        
        # Get enrollment count
        count = await db.enrollments.count_documents({
            'class_id': class_id,
            'status': 'active'
        })
        class_item['enrolled_count'] = count
        
        return Class(**class_item)
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.put("/{class_id}", response_model=Class)
async def update_class(class_id: str, class_update: ClassUpdate):
    """Update a class"""
    try:
        existing = await db.classes.find_one({"id": class_id})
        if not existing:
            raise HTTPException(status_code=404, detail="Class not found")
        
        update_data = class_update.dict(exclude_unset=True)
        update_data['updated_at'] = datetime.utcnow()
        
        await db.classes.update_one(
            {"id": class_id},
            {"$set": update_data}
        )
        
        updated = await db.classes.find_one({"id": class_id})
        
        # Get enrollment count
        count = await db.enrollments.count_documents({
            'class_id': class_id,
            'status': 'active'
        })
        updated['enrolled_count'] = count
        
        return Class(**updated)
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.delete("/{class_id}")
async def delete_class(class_id: str):
    """Delete a class"""
    try:
        result = await db.classes.delete_one({"id": class_id})
        if result.deleted_count == 0:
            raise HTTPException(status_code=404, detail="Class not found")
        return {"message": "Class deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

# Enrollment endpoints
@router.post("/{class_id}/enroll")
async def enroll_student(
    class_id: str,
    student_id: str = Query(...),
    student_name: str = Query(...)
):
    """Enroll a student in a class"""
    try:
        # Check if class exists
        class_item = await db.classes.find_one({"id": class_id})
        if not class_item:
            raise HTTPException(status_code=404, detail="Class not found")
        
        # Check if already enrolled
        existing = await db.enrollments.find_one({
            'class_id': class_id,
            'student_id': student_id,
            'status': 'active'
        })
        if existing:
            raise HTTPException(status_code=400, detail="Student already enrolled in this class")
        
        # Check capacity
        enrolled_count = await db.enrollments.count_documents({
            'class_id': class_id,
            'status': 'active'
        })
        if enrolled_count >= class_item.get('capacity', 20):
            raise HTTPException(status_code=400, detail="Class is full")
        
        # Create enrollment
        enrollment = EnrollmentRecord(
            class_id=class_id,
            student_id=student_id,
            student_name=student_name
        )
        
        await db.enrollments.insert_one(enrollment.dict())
        return enrollment
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.delete("/{class_id}/enroll/{student_id}")
async def unenroll_student(class_id: str, student_id: str):
    """Remove a student from a class"""
    try:
        result = await db.enrollments.update_one(
            {
                'class_id': class_id,
                'student_id': student_id,
                'status': 'active'
            },
            {'$set': {'status': 'dropped'}}
        )
        
        if result.modified_count == 0:
            raise HTTPException(status_code=404, detail="Enrollment not found")
        
        return {"message": "Student unenrolled successfully"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/{class_id}/students")
async def get_class_students(class_id: str):
    """Get all students enrolled in a class"""
    try:
        enrollments = await db.enrollments.find({
            'class_id': class_id,
            'status': 'active'
        }).to_list(500)
        
        # Get full student details
        student_ids = [e['student_id'] for e in enrollments]
        students = await db.students.find({
            'id': {'$in': student_ids}
        }).to_list(500)
        
        # Combine enrollment and student data
        result = []
        for enrollment in enrollments:
            student = next((s for s in students if s['id'] == enrollment['student_id']), None)
            if student:
                result.append({
                    'enrollment_id': enrollment['id'],
                    'enrollment_date': enrollment['enrollment_date'],
                    'student': student
                })
        
        return result
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/stats/{school_id}")
async def get_class_stats(school_id: str):
    """Get class statistics for a school"""
    try:
        total = await db.classes.count_documents({"school_id": school_id})
        active = await db.classes.count_documents({"school_id": school_id, "status": "active"})
        full = await db.classes.count_documents({"school_id": school_id, "status": "full"})
        
        # Count by sport
        pipeline = [
            {"$match": {"school_id": school_id}},
            {"$group": {"_id": "$sport", "count": {"$sum": 1}}}
        ]
        sports = await db.classes.aggregate(pipeline).to_list(100)
        
        # Total enrolled students across all classes
        total_enrolled = await db.enrollments.count_documents({
            "status": "active"
        })
        
        return {
            "total": total,
            "active": active,
            "full": full,
            "by_sport": {s['_id']: s['count'] for s in sports if s['_id']},
            "total_enrolled": total_enrolled
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
